### PR TITLE
Don't mutate frozen strings

### PR DESCRIPTION
**Pull request by jfly**

### CODE REVIEW

File: WcaOnRails/spec/factories/persons.rb [hunk 1]
```diff
@@ -4,7 +4,10 @@
     wca_id do
       mid = ('A'..'Z').to_a.sample(4).join
       "2016#{mid}01".tap do |id|
-        id.next! while Person.exists?(wca_id: id)
+        while Person.exists?(wca_id: id)
+          id = id.next
+        end
+        id
```
Comment (jonatanklosko) [inline WcaOnRails/spec/factories/persons.rb]: You don't have to explicitly return id here, that's exactly what the `tap` method is for =)

Comment (jfly) [inline WcaOnRails/spec/factories/persons.rb]: Ah, good catch! If we don't mutate the string, then this `tap` is pointless. I've rewritten this.

File: WcaOnRails/spec/factories/persons.rb [hunk 2]
```diff
@@ -3,9 +3,11 @@
   factory :person do
     wca_id do
       mid = ('A'..'Z').to_a.sample(4).join
-      "2016#{mid}01".tap do |id|
-        id.next! while Person.exists?(wca_id: id)
+      id = "2016#{mid}01"
+      while Person.exists?(wca_id: id)
+        id = id.next
       end
```
Comment (jonatanklosko) [inline WcaOnRails/spec/factories/persons.rb]: Rubocop dislikes this ([=)](https://github.com/bbatsov/ruby-style-guide#while-as-a-modifier)).

File: WcaOnRails/spec/factories/persons.rb [hunk 3]
```diff
@@ -3,9 +3,9 @@
   factory :person do
     wca_id do
       mid = ('A'..'Z').to_a.sample(4).join
-      "2016#{mid}01".tap do |id|
-        id.next! while Person.exists?(wca_id: id)
-      end
+      id = "2016#{mid}01"
+      id = id.next while Person.exists?(wca_id: id)
+      id
```
Comment (larspetrus) [inline WcaOnRails/spec/factories/persons.rb]: Not sure it's an improvement, but this is as far as I can squeeze this one:

    wca_id do
      mid = ('A'..'Z').to_a.sample(4).join
      num = 0
      begin id = "2016#{mid}0#{num += 1}" end while Person.exists?(wca_id: id)
      id
    end

Comment (jfly) [inline WcaOnRails/spec/factories/persons.rb]: I had to leave the explicit return `id` at the end because line 7 (with the inline while loop) returns `nil`:

```
irb(main):006:0> i = 1
=> 1
irb(main):007:0> i = 3 while i == 1
=> nil
irb(main):008:0> i
=> 3
```

@larspetrus, I fear your solution becomes a problem once num equals 99, whereas `.next` will do a cool thing:

```
irb(main):009:0> "2016FLEI99".next
=> "2016FLEJ00"
```

Comment (larspetrus) [inline WcaOnRails/spec/factories/persons.rb]: I would personally ignore such an edge case in test code, but your 4 lines are at least as good as my four lines, so it doesn't matter.

I'm slightly frustrated that `foo = bar() while maybe(foo)` returns nil rather than `foo`. I suspect Ruby has a better construct like this, but I'm not spending any more time on this piece of bikeshed trivia this morning.

**Comment by jonatanklosko** [comment]
LGTM =)

**Comment by jfly** [comment]
Thanks! Merging up now.
